### DISCUSSION TITLE
Shared folder on RedHat guest

### DISCUSSION THREAD

**Discussion by lelegard** [utmapp/UTM#5402] [Linux] [answered]
Hi,

How do you mount shared folders on RedHat guests?

It seems that RedHat does not support 9pfs.

I already have Ubuntu, Debian and Windows 11 virtual machines on a M1 Mac. On Ubuntu and Debian, the shared folder is mounted using the 9p file system. On Windows, it uses the WebDAV equivalent.

After installing a RedHat virtual machine (Rocky Linux 9.2), I realize that 9p does not seem to be supported:

~~~
$ sudo mount -t 9p share /shared
mount: /shared: unknown filesystem type '9p'.
$
$ grep -i 9p /boot/config-5.14.0-284.11.1.el9_2.aarch64 
# CONFIG_NET_9P is not set
$
$ ls /lib/modules/5.14.0-284.11.1.el9_2.aarch64/kernel/fs
afs                ceph   ext4     fuse   lockd          nfs         nls        smbfs_common  xfs
binfmt_misc.ko.xz  cifs   fat      isofs  mbcache.ko.xz  nfs_common  overlayfs  squashfs
cachefiles         exfat  fscache  jbd2   netfs          nfsd        pstore     udf
$
$ rpm -qa '*spice*'
spice-vdagent-0.21.0-5.el9.aarch64
~~~

Recompiling the kernel is pretty odd in well-packaged Entreprise-class distros like RedHat. It can be done, but if you need to do so, you will probably use another distro (Arch, Gentoo, whatever).

I tried to use `virtiofs` without luck (using libvirt 9.0):
~~~
$ sudo mount -t virtiofs share /shared
mount: /shared: wrong fs type, bad option, bad superblock on share, missing codepage or helper program, or other error.
~~~

For the comparison, on an Ubuntu 23.04 VM:
~~~
$ grep -i 9p /boot/config-6.2.0-20-generic 
CONFIG_NET_9P=m
CONFIG_NET_9P_FD=m
CONFIG_NET_9P_VIRTIO=m
CONFIG_NET_9P_XEN=m
CONFIG_NET_9P_RDMA=m
# CONFIG_NET_9P_DEBUG is not set
CONFIG_VIDEO_MT9P031=m
CONFIG_9P_FS=m
CONFIG_9P_FSCACHE=y
CONFIG_9P_FS_POSIX_ACL=y
CONFIG_9P_FS_SECURITY=y
$ 
$ ls /lib/modules/6.2.0-20-generic/kernel/fs
9p	befs		ceph	efs    freevxfs  hfsplus  ksmbd  nfs_common  ntfs3	pstore	  romfs		udf
adfs	bfs		cifs	erofs  fscache	 hpfs	  lockd  nfsd	     ocfs2	qnx4	  shiftfs.ko	ufs
affs	binfmt_misc.ko	coda	exfat  fuse	 isofs	  minix  nilfs2      omfs	qnx6	  smbfs_common	xfs
afs	btrfs		cramfs	f2fs   gfs2	 jffs2	  netfs  nls	     orangefs	quota	  sysv		zonefs
autofs	cachefiles	dlm	fat    hfs	 jfs	  nfs	 ntfs	     overlayfs	reiserfs  ubifs
$
$ dpkg -l '*spice*'
[...]
ii  spice-vdagent  0.22.1-3     arm64        Spice agent for Linux
~~~

**Comment by dmichael89** [comment, marked as answer]
You could give the elrepo-kernel from the ELRepo.org Community Enterprise Linux Kernel Repository a try.

**Comment by lelegard** [comment]
Thanks @dmichael89 

> elrepo-kernel from the ELRepo.org Community Enterprise Linux Kernel Repository

It works, I got a `9p` kernel module and I can mount UTM shared folders. However, it significantly changes the RH9 environment. So, this is fine if you really want to have access to the shared folders. It is not so fine if you want to test a true RH9 environment.

- The kernel which comes with RH9 is (currently) 5.14.
- The kernel which comes with `elrepo-kernel` is one of 
  - 6.1 (`kernel-lt` package for long-term)
  - 6.4 (`kernel-ml` package for mainline).

Technically, it works. As a reference environment, this is no longer a RH9 setup.

For reference:
~~~
rpm --import https://www.elrepo.org/RPM-GPG-KEY-elrepo.org
dnf install https://www.elrepo.org/elrepo-release-9.el9.elrepo.noarch.rpm
dnf config-manager --set-enabled elrepo-kernel
dnf install kernel-ml
reboot
~~~
Then, you can mount `9p` file systems (UTM shared folders).
